### PR TITLE
feat/firestore-rules-tests

### PR DESCRIPTION
**Pull request by GuyAvraham**

### CODE REVIEW

**Comment by rootux** [comment]
Seems like tests are still failing.

**Comment by GuyAvraham** [comment]
This PR is not ready yet.. waiting for circle CI support

File: .circleci/config.yml [hunk 1]
```diff
@@ -3,7 +3,11 @@ version: 2.0
 jobs:
   build:
     docker:
-      - image: circleci/node:8.11
+      - image: circleci/node:8.15.1
```
Comment (GuyAvraham) [inline .circleci/config.yml:6]: for java 8 to work

File: .circleci/config.yml [hunk 2]
```diff
@@ -27,6 +31,14 @@ jobs:
             - node_modules
           key: v1-dependencies-{{ checksum "package.json" }}
 
+      - run:
+          name: Test firestore rules
+          command: |
+            sudo apt-get update
+            sudo apt install openjdk-8-jdk
```
Comment (GuyAvraham) [inline .circleci/config.yml:38]: not sure how to cache it

File: rules/firestore.rules [hunk 3]
```diff
@@ -50,14 +50,8 @@ service cloud.firestore {
...
 
-    // Since firebase doesn't have the ability to specify what CAN be updated
-    // We need to say what CANNOT be updated - that is all of the tasks fields BUT listeners
     function isNotUpdatingAnythingButListeners() {
-      return notUpdating('id') && notUpdating('title') && notUpdating('assignee')
-      && notUpdating('type') && notUpdating('projectName') && notUpdating('label')
-      && notUpdating('created') && notUpdating('dueDate') && notUpdating('creator')
-      && notUpdating('isDone') && notUpdating('doneDate') && notUpdating('description')
-      && notUpdating('requirements') && notUpdating('isCritical') && notUpdating('extraFields');
+      return request.resource.data.size() == 1 && request.resource.data.keys().hasOnly(['listeners'])
```
Comment (GuyAvraham) [inline rules/firestore.rules]: @rootux let me know if you agree here. the tests are good with that

File: rules/firestore.rules [hunk 4]
```diff
@@ -76,8 +70,10 @@ service cloud.firestore {
         // Uncomment the following to allow unassigned task to be assigned by all
         allow update: if isSignedIn() && isNotUpdatingAnythingButListeners()
             ||
-              (isUserInvited(anyProject) && (isGuide() || isUserAssigneeOrCreator()
-              || isTaskUnassigned() || isNotUpdatingAnythingButListeners())
+              (
+                isUserInvited(anyProject) && (
+                  isGuide() || isUserAssigneeOrCreator() || isTaskUnassigned() || isNotUpdatingAnythingButListeners()
+                )
```
Comment (GuyAvraham) [inline rules/firestore.rules:80]: formatting

File: .circleci/config.yml (hunk 2 above)
Comment (rootux) [reply .circleci/config.yml:38]: That's okay for now

File: rules/firestore.rules (hunk 3 above)
Comment (rootux) [reply rules/firestore.rules]: No. This is not good. Please read the stack overflow link I've commented just above that line.t. This was the old solution we used to have. It's not working

Comment (rootux) [reply rules/firestore.rules]: Read the comment under notUpdating function

Comment (GuyAvraham) [reply rules/firestore.rules]: 1. I understand and will revert.
2. I'm curious, if it is the case and request.size should be as the next document state (having all the existing fields) how come on  the test of the listeners, the test passes when I send only listeners, check that the size of the request is 1, but the existing doc is bigger then one field (u can see my console log on line 384 if u fetch the branch)

Comment (GuyAvraham) [reply rules/firestore.rules]: the tests pass by this and by your implementation.
@rootux I want to be able to write a test that will fail the wrong implementation

Comment (rootux) [reply rules/firestore.rules]: As for 2. That might be due to different behavior on emulator and production.
I've talked with the firebase team and they are working on a proper solution

Comment (GuyAvraham) [reply rules/firestore.rules]: wow amazing! can u frw me the email with their response

**Comment by rootux** [comment]
Can't remember the thread but it was either on their github issues or on
stackoverflow.
I remember it involved frank -
https://stackoverflow.com/users/209103/frank-van-puffelen

**Pull request merged**